### PR TITLE
Initial implementation of Baron Bunny as a gadget

### LINKED ISSUE #1282: Amber's Baron Bunny should be implemented as a Gadget
Amber's Baron Bunny:
- can only hold Cryo durability in normal circumstances 
- this can be picked up by absorb checks
- properties from Dendro MC's Q gadget also apply here: 
  - not able to proc OnReaction affects on Baron Bunny
  - ~~only Sucrose E and Faruzan Pressurized Collapse can interact with Baron Bunny for a Swirl (don't forget to set the durability to 50 for the incoming anemo attack)~~
  - once it has Cryo durability, any elemental application except Anemo (see above) and Dendro (because it has no reaction with Cryo) can interact with it for a reaction
  - ...

https://library.keqingmains.com/evidence/characters/pyro/amber#baron-bunnys-aura

### PR DESCRIPTION
**Pull request by Charlie-Zheng**
Closes #1282

### CODE REVIEW

**Comment by Charlie-Zheng** [comment]
Note: I found no sources showing that Baron Bunny interactions with Sucrose/Faruzan skill are 50 gauge? I already found some inaccuracies in the original issue, so I'm awaiting further confirmation of this feature

File: internal/characters/amber/bunny.go [hunk 1]
```diff
@@ -108,12 +236,11 @@ func (c *char) overloadExplode() {
 		if atk.Info.AttackTag != attacks.AttackTagOverloadDamage {
 			return false
 		}
+		c.bunnies[0].ae.Info.Abil = manualExplosionAbil
```
Comment (k0l11) [inline internal/characters/amber/bunny.go:239]: Shouldn't this be set for all the bunnies exploded here? So do like `v.ae.Info.Abil = manualExplosionAbil` as part of the loop below.

File: internal/characters/amber/bunny.go [hunk 2]
```diff
@@ -108,12 +236,11 @@ func (c *char) overloadExplode() {
...
 
 		for _, v := range c.bunnies {
-			c.bunnies[0].ae.Info.Abil = manualExplosionAbil
-			c.Core.QueueAttackEvent(&v.ae, 1)
+			v.Kill()
 		}
```
Comment (k0l11) [inline internal/characters/amber/bunny.go:243]: Please check whether the current bunny is within the overload attack pattern via `v.IsWithinArea(atk.Pattern)` before exploding it. If the bunny is out of range, it shouldn't explode.

File: internal/characters/amber/bunny.go [hunk 3]
```diff
@@ -1,25 +1,159 @@
...
+		r.TryMelt(a)
+	case attributes.Cryo:
+	case attributes.Hydro:
+		r.TryFreeze(a)
+	case attributes.Anemo:
+		r.TrySwirlCryo(a)
+		r.TrySwirlFrozen(a)
+	case attributes.Geo:
+		r.TryCrystallizeCryo(a)
+		r.TryCrystallizeFrozen(a)
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/amber/bunny.go:124]: You can apply cryo, apply hydro, then apply anemo

File: internal/characters/amber/bunny.go (hunk 1 above)
Comment (Charlie-Zheng) [reply internal/characters/amber/bunny.go:239]: I'm not sure. The original code did this and there is no evidence to the contrary so I didn't change it.

File: internal/characters/amber/bunny.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (k0l11) [reply internal/characters/amber/bunny.go:124]: How are you applying hydro if hydro atks coming from the char doesn't do anything? For example, Kaeya E -> Childe E -> Zhongli E doesn't shatter Baron Bunny. But if you replace Childe E with overworld water it does shatter.

File: internal/characters/amber/bunny.go (hunk 1 above)
Comment (k0l11) [reply internal/characters/amber/bunny.go:239]: I think it should be set for all bunnies.

File: internal/characters/amber/bunny.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (k0l11) [reply internal/characters/amber/bunny.go:124]: [TCL wording](https://library.keqingmains.com/evidence/characters/pyro/amber#baron-bunnys-aura) is wonky on this, but yes you're right.

File: internal/characters/amber/bunny.go [hunk 4]
```diff
@@ -1,25 +1,159 @@
...
+		r.TryFrozenSuperconduct(a)
+		r.TrySuperconduct(a)
+	case attributes.Pyro:
+		r.TryMelt(a)
+	case attributes.Cryo:
+	case attributes.Hydro:
+		r.TryFreeze(a)
+	case attributes.Anemo:
+		r.TrySwirlCryo(a)
+		r.TrySwirlFrozen(a)
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/amber/bunny.go:121]: I just checked in game it looks like the frozen aura just does not attach. But the reaction needs to occur to reduce cryo aura.

File: internal/characters/amber/bunny.go [hunk 5]
```diff
@@ -1,25 +1,159 @@
...
-	b := bunny{}
-	b.src = c.Core.F
+func (r *Bunny) React(a *combat.AttackEvent) {
+	//only check the ones possible
+	switch a.Info.Element {
+	case attributes.Electro:
+		r.TryFrozenSuperconduct(a)
+		r.TrySuperconduct(a)
+	case attributes.Pyro:
+		r.TryMelt(a)
```
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/amber/bunny.go:115]: I'm pretty sure it reduces the cryo aura

[the remark this answers is not in the document]
Comment (k0l11) [reply internal/characters/amber/bunny.go:115]: Alright

File: internal/characters/amber/bunny.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Charlie-Zheng) [reply internal/characters/amber/bunny.go:124]: Looks like the frozen reaction just does not occur.

File: internal/characters/amber/bunny.go (hunk 2 above)
Comment (Charlie-Zheng) [reply internal/characters/amber/bunny.go:243]: Is this supposed to explode the bunnies if it's the burst that triggers the OL?

Comment (k0l11) [reply internal/characters/amber/bunny.go]: No, just when the CA causes overloaded. Just checked ingame.

Comment (Charlie-Zheng) [reply internal/characters/amber/bunny.go]: How do I do the `v.IsWithinArea(atk.Pattern)` , the atk given by the OL subscription is the triggering atk. Do I just do my own  `combat.NewCircleHitOnTarget(target, nil, 3)`?

**Review by k0l11 — APPROVED** [review]

**Pull request merged**